tinygrad/runtime/ops_metal.py

```diff
@@ -65,19 +68,28 @@ def metal_src_to_library(device:MetalDevice, src:str) -> objc_instance:
   return library
 
 class MetalCompiler(Compiler):
-  def __init__(self, dev:Optional[MetalDevice]=None):
-    self.dev = dev
-    super().__init__("compile_metal_xcode" if self.dev is None else "compile_metal")
+  def __init__(self):
+    if not getenv("METAL_XCODE", 0): self.cgs = compiler.MTLCodeGenServiceCreate(b"tinygrad")
+    super().__init__("compile_metal_xcode" if getenv("METAL_XCODE", 0) else "compile_metal")
+  def __reduce__(self): return (MetalCompiler,()) # to create new instance for each multiprocessing fork
   def compile(self, src:str) -> bytes:
-    if self.dev is None:
+    if self.cgs is None:
       # NOTE: if you run llvm-dis on "air" you can see the llvm bytecode
       air = subprocess.check_output(['xcrun', '-sdk', 'macosx', 'metal', '-x', 'metal', '-c', '-', '-o', '-'], input=src.encode('utf-8'))
       lib = subprocess.check_output(['xcrun', '-sdk', 'macosx', 'metallib', '-', '-o', '-'], input=air)
     else:
-      library = metal_src_to_library(self.dev, src)
-      library_contents = msg(library, "libraryDataContents", restype=objc_instance)
-      lib = ctypes.string_at(msg(library_contents, "bytes"), cast(int, msg(library_contents, "length", restype=ctypes.c_ulong)))
-    assert lib[:4] == b"MTLB", "Invalid Metal library. Using conda? Corrupt XCode?"
+      lib, err = None, CompileError("MTLCodeGenServiceBuildRequest returned without calling the callback")
+      @ctypes.CFUNCTYPE(None, ctypes.c_void_p, ctypes.c_int32, ctypes.c_void_p, ctypes.c_size_t, ctypes.c_char_p)
+      def callback(blockptr, error, dataPtr, dataLen, errorMessage):
+        nonlocal lib, err
+        if error == 0: lib = (reply:=bytes(to_mv(dataPtr, dataLen)))[sum(struct.unpack('<LL', reply[8:16])):]
+        else: err = CompileError(errorMessage.decode())
+      params = f'-fno-fast-math -std=metal3.1 --driver-mode=metal -x metal -fmodules-cache-path=\"{os.path.join(_cache_dir, "tinygrad")}\"'
```

Comment:
Do you need a cache?

Reply:
Yes, otherwise it's 250 ms instead of 8 ms. As i understand it caches stdlib compilation

Reply:
No cache:
Compilation took 254.2 ms
Cache:
Compilation took 7.8 ms

This was in compile_direct.py

Reply:
There should be a comment explaining this, right? It's not obvious why.

Saying "this was in compile_direct.py" does nothing for me.